src/x86_64/page_machine.h

```diff
@@ -0,0 +1,297 @@
+{
+    return (pageflags){.w = flags.w & ~PAGE_NO_EXEC};
+}
+
+static inline pageflags pageflags_minpage(pageflags flags)
+{
+    return (pageflags){.w = flags.w | PAGE_NO_PS};
+}
+
+static inline pageflags pageflags_no_minpage(pageflags flags)
+{
+    return (pageflags){.w = flags.w & ~PAGE_NO_PS};
+}
+
+/* no-exec, read-only */
+static inline pageflags pageflags_default_user(void)
+{
+    return pageflags_user(pageflags_minpage(pageflags_memory()));
+}
+
+static inline boolean pageflags_is_writable(pageflags flags)
+{
+    return (flags.w & PAGE_WRITABLE) != 0;
+}
+
+static inline boolean pageflags_is_readonly(pageflags flags)
+{
+    return !pageflags_is_writable(flags);
+}
+
+static inline boolean pageflags_is_noexec(pageflags flags)
+{
+    return (flags.w & PAGE_NO_EXEC) != 0;
+}
+
+static inline boolean pageflags_is_exec(pageflags flags)
+{
+    return !pageflags_is_noexec(flags);
+}
+
+typedef u64 pte;
+typedef volatile pte *pteptr;
+
+static inline pte pte_from_pteptr(pteptr pp)
+{
+    return *pp;
+}
+
+static inline void pte_set(pteptr pp, pte p)
+{
+    *pp = p;
+}
+
+static inline boolean pte_is_present(pte entry)
+{
+    return (entry & PAGE_PRESENT) != 0;
+}
+
+static inline boolean pte_is_block_mapping(pte entry)
+{
+    return (entry & PAGE_PS) != 0;
+}
+
+static inline int pt_level_shift(int level)
+{
+    switch (level) {
+    case 1:
+        return PT_SHIFT_L1;
+    case 2:
+        return PT_SHIFT_L2;
+    case 3:
+        return PT_SHIFT_L3;
+    case 4:
+        return PT_SHIFT_L4;
+    }
+    return 0;
+}
+
+static inline u64 flags_from_pte(u64 pte)
+{
+    return pte & PAGE_FLAGS_MASK;
+}
+
+static inline u64 page_pte(u64 phys, u64 flags)
+{
+    return phys | (flags & ~PAGE_NO_PS) | PAGE_PRESENT;
+}
+
+static inline u64 block_pte(u64 phys, u64 flags)
+{
+    return phys | flags | PAGE_PRESENT | PAGE_PS;
+}
+
+static inline u64 new_level_pte(u64 tp_phys)
+{
+    return tp_phys | PAGE_WRITABLE | PAGE_USER | PAGE_PRESENT;
+}
+
+static inline boolean flags_has_minpage(u64 flags)
+{
+    return (flags & PAGE_NO_PS) != 0;
+}
+
+static inline u64 canonize_address(u64 addr)
+{
+    if (addr & U64_FROM_BIT(47))
+        addr |= 0xffff000000000000;
+    return addr;
+}
+
+extern u64 pagebase;
+static inline u64 get_pagetable_base(u64 vaddr)
+{
+    return pagebase;
+}
+
+u64 *pointer_from_pteaddr(u64 pa);
+
+/* log of mapping size (block or page) if valid leaf, else 0 */
+static inline int pte_order(int level, pte entry)
+{
+    assert(level > 0 && level < 5);
+    if (level == 1 || !pte_is_present(entry) ||
+        (level != 4 && !(entry & PAGE_PS)))
+        return 0;
+    return pt_level_shift(level);
+}
+
+static inline u64 pte_map_size(int level, pte entry)
+{
+    if (pte_is_present(entry)) {
+        // XXX take from arm
+        if (level == 4)
+            return PAGESIZE;
+        if ((entry & PAGE_PS) && level != 1)
+            return level == 2 ? HUGE_PAGESIZE : PAGESIZE_2M;
```

Comment:
HUGE_PAGESIZE is defined as 0x100000000ull (4 GB), this function should return 1 GB instead.